Initial Multichain MIP drafts (#31)
MetaMask/metamask-improvement-proposals: MIPs/mip-mc-request.md

## Motivation
Both users and dapp builders are increasingly interested in experiences that involve multiple chains. MetaMask shouldn't get in their way.

This proposal is intended to support the following objectives:
- Unlocking user experiences that involve minimal friction when interacting across different chains (both in dApps and MetaMask).
- Simplifying development for dapp builders.
- Aligning MetaMask with community standards to simplify integrations and tooling in the ecosystem.
- Increasing MetaMask's utility as a universal wallet solution.

Improved conversion rates on current activities that involve multiple chains can be realized through the Multichain API. More importantly, the implementation of a multichain API will serve as a significant enabler of permissionless innovation in the community. We anticipate that it will drive dramatic growth in existing and entirely new categories of dApps.

# Proposal

## Language
The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written in uppercase in this document are to be interpreted as described in RFC 2119.

## Definitions

> Add a note about how we use the terms `chain` and `network` interchangeably

Reviewer: haha this is an "ethereumism" and some non-chain networks like DAGs might feel more welcome in the Snaps project if you discriminate!

    "params":[{
        "version": "0.0.1",
        "scope": "eip155:1",
        "request":{
            "method": "eth_getBalance",
            "params": ["0xabc...","latest"]
        }
    }]
    "id": 1
});
```

In this example, the `scope` field is populated with the chain ID corresponding to Ethereum Mainnet (as per CAIP-10), and the `request` field contains the typical JSON-RPC request for getting an Ethereum balance.

##### Batch Request for Multiple Chains
Here's an example of how JSON-RPC batch requests to get balances from both Ethereum Mainnet and Binance Smart Chain would be called by wrapping the requests in a `provider_request`:

```javascript
ethereum.request([
  {
    "jsonrpc": "2.0",
    "method": "provider_request",
    "params": {
        "scope": "eip155:1",
        "request": {
            "method": "eth_getBalance",
            "params": ["0xabc...", "latest"]
      }
    }
    "id": 1
  },
  {
    "jsonrpc": "2.0",
    "method": "provider_request",
    "params": {
        "scope": "eip155:59144",
        "request": {
            "method": "eth_getBalance",
            "params": ["0xabc...", "latest"]
      }
    }
    "id": 2
  }
]);
```

> In this example, the `provider_request` wraps each JSON-RPC `eth_getBalance` request with one scoped to Ethereum Mainnet (`eip155:1`) and another to Linea (`eip155:59144`).

Reviewer: ```suggestion
In this example, the `provider_request` wraps each JSON-RPC `eth_getBalance` request with a separate CAIP-27 request, one explicitly scoped to Ethereum Mainnet (`eip155:1`) and another to Linea (`eip155:59144`).
```

> For consistency and excepting [provider_authorization](mip-mc-handshake.md), all MetaMask JSON-RPC requests will be made through a `provider_request` including those that are `wallet` scoped. See [provider_authorization](mip-mc-handshake.md) for information about supported scopes.

Reviewer 2: ```suggestion
To maintain consistency, all MetaMask JSON-RPC requests except [provider_authorization](mip-mc-handshake.md)will be made through `provider_request` including those that are `wallet` scoped. See [provider_authorization](mip-mc-handshake.md) for information about supported scopes.
```

## Test Cases
General test cases should cover:
- Successful request with correct `scope`
- Error handling when `scope` is not provided
- Error handling when an unsupported `scope` is provided

Test cases for specific JSON-RPC `method` calls are outside the scope of this proposal.

## Caveats
[List any potential drawbacks, limitations, or risks associated with the proposal]

## Implementation
MetaMask's JSON-RPC request-handling code needs to be updated to support this new `provider_request` method. Appropriate UI updates are also needed to ensure that users are sufficiently informed and aware of what they are interacting with or being asked to confirm.

A new suite of tests will be added to comprehensively cover key test cases.
Technical documentation and developer tools will have to be updated to include the Multichain API.

API Maintainers will be implementing the Multichain API in coordination with other MetaMask team members.

## Developer Adoption Considerations
[Explain any considerations that developers should take into account when adopting this proposal. For example, how will it affect compatibility, and what changes may need to be made to accommodate the proposal?]

> The implementation would be designed to be fully backward-compatible. Existing methods for changing networks and executing transactions would remain unaffected. Once there is sufficient industry adoption of the Multichain API, backward-compatibility may be reconsidered and gradually deprecated. 

Reviewer: It might help to refer specifically to CAIP-25 adoption here-- CAIP-27 is supposed to work even if CAIP-25 isn't adopted (or perfectly implemented), i.e., a "scope" in CAIP-27 *should* correspond to a scopeObject authorized with the handshake, but conceivably some wallets might implement 27 without implementing 25 as a "halfmeasure" or a way of working their way up to the more stateful CAIP-25 concept of scopes by just handling these strings as an interim goal...
